src/hmem_cuda_gdrcopy.c

```diff
@@ -0,0 +1,380 @@
+		global_gdrcopy_ops.gdr_close(global_gdr);
+		cuda_gdrcopy_dl_hmem_cleanup();
+		return err;
+	}
+
+	return FI_SUCCESS;
+}
+
+int cuda_gdrcopy_hmem_cleanup(void)
+{
+	int err, ret = 0;
+
+	err = pthread_spin_destroy(&global_gdr_lock);
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "destroy global_gdr_lock failed! err: %s\n",
+			strerror(err));
+		ret = err;
+	}
+
+	assert(global_gdrcopy_ops.gdr_close);
+	err = global_gdrcopy_ops.gdr_close(global_gdr);
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "close global_gdr failed! err: %s\n",
+			strerror(err));
+		ret = err;
+	}
+
+	err = cuda_gdrcopy_dl_hmem_cleanup();
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "cuda_gdrcopy_dl_hmem_cleaup() failed! err: %s\n",
+			strerror(err));
+		ret = err;
+	}
+
+	return ret;
+}
+
+int cuda_gdrcopy_to_dev(uint64_t handle, void *devptr, const void *hostptr, size_t len)
+{
+	ssize_t off;
+	struct ofi_gdrcopy_handle *gdrcopy;
+	void *gdrcopy_user_ptr;
+
+	assert(global_gdrcopy_ops.gdr_copy_to_mapping);
+	assert(handle);
+
+	gdrcopy = (struct ofi_gdrcopy_handle *)handle;
+	off = (char *)devptr - (char *)gdrcopy->cuda_ptr;
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_to_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return 0;
+}
+
+int cuda_gdrcopy_from_dev(uint64_t handle, void *hostptr, const void *devptr, size_t len)
+{
+	ssize_t off;
+	struct ofi_gdrcopy_handle *gdrcopy;
+	void *gdrcopy_user_ptr;
+
+	assert(global_gdrcopy_ops.gdr_copy_from_mapping);
+
+	gdrcopy = (struct ofi_gdrcopy_handle *)handle;
+	off = (char *)devptr - (char *)gdrcopy->cuda_ptr;
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_from_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return FI_SUCCESS;
+}
+
+int cuda_gdrcopy_dev_register(struct fi_mr_attr *mr_attr, uint64_t *handle)
+{
+	int err;
+	uintptr_t regbgn = (uintptr_t)mr_attr->mr_iov->iov_base;
+	uintptr_t regend = (uintptr_t)mr_attr->mr_iov->iov_base + mr_attr->mr_iov->iov_len;
+	size_t reglen;
+	struct ofi_gdrcopy_handle *gdrcopy;
+
+	assert(global_gdr);
+	assert(global_gdrcopy_ops.gdr_pin_buffer);
+	assert(global_gdrcopy_ops.gdr_map);
+
+	regbgn = regbgn & GPU_PAGE_MASK;
+	regend = (regend & GPU_PAGE_MASK) + GPU_PAGE_SIZE;
+	reglen = regend - regbgn;
+
+	gdrcopy = malloc(sizeof(struct ofi_gdrcopy_handle));
+	if (!gdrcopy)
+		return -FI_ENOMEM;
+
+	assert(global_gdr);
+	pthread_spin_lock(&global_gdr_lock);
+	err = global_gdrcopy_ops.gdr_pin_buffer(global_gdr, regbgn,
+					 reglen, 0, 0, &gdrcopy->mh);
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "gdr_pin_buffer failed! error: %s",
+			strerror(err));
+		free(gdrcopy);
+		goto exit;
+	}
+
+	gdrcopy->cuda_ptr = (void *)regbgn;
+	gdrcopy->length = reglen;
+
+	err = global_gdrcopy_ops.gdr_map(global_gdr, gdrcopy->mh, &gdrcopy->user_ptr, gdrcopy->length);
```

Comment:
I'm only picking one spot, but please check line lengths throughout the patch and try to limit to roughly 80 characters.

Reply:
Fixed in new revision.